File: 797. All Paths From Source to Target/AllPathsSourceTarget.py
'''
https://leetcode.com/problems/all-paths-from-source-to-target/
'''


from typing import List


class Solution:
    # O(n^2) time,
    # O(n^2) space,
    # Approach: DFS
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        n = len(graph)
        paths = []
        
        def dfs(root:int, path: List[int]) -> None:
            if root == n-1:
                paths.append(path[::])
                
            children = graph[root]
            
            for child in children:
                path.append(child)
                dfs(child, path)
                path.pop()
                
        dfs(0, [0])
        
        return paths

File: 797. All Paths From Source to Target/allPathsSourceTarget.go
package allPathsSourceTarget


// O(n+e) time,
// O(n) space,
// Approach: dfs, backtracking, 
func traverse(node int, path *[]int, graph *[][]int, pathes *[][]int) {
    *path = append(*path, node)
    if node == len(*graph)-1 {
        final_path := make([]int, len(*path))
        copy(final_path, *path)
        (*pathes) = append(*pathes, final_path)
        *path = (*path)[:len(*path)-1]
        return
    }
    
    for _, nbr := range (*graph)[node] {
        traverse(nbr, path, graph, pathes)
    }
    *path = (*path)[:len(*path)-1]
}


func allPathsSourceTarget(graph [][]int) [][]int {
    var all_pathes [][]int
    var path []int
    traverse(0, &path, &graph, &all_pathes)
    
    return all_pathes
}